Use Google Map for choosing Image Set Location

iezer opened the pull request:
@bantic  For you CC @redyaffle 

Adds a Google Map for setting image set location. 

The latest seeds.db on the api branch seeds the lat/long for image sets which makes this easier to test.

Here's what it looks like

![](http://g.recordit.co/25LB6MXNpM.gif)

iezer committed:
Use Google Map for choosing Image Set Location

iezer commented on app/routes/image-set.js:
```diff
@@ -3,6 +3,17 @@ import OrganizationRouteMixin from 'lion-guardians/mixins/organization-route';
+      if (window.SKIP_MAPS) {
+        return Ember.RSVP.resolve();
+      } else {
+        return route.loadGoogleMap();
```
@bantic  This lazy loads the Google Map only when we enter an image-set route. However I want to push this down further and only load when a user clicks the edit button on an image-set. The plugin only injects the loadGoogleMap() function on routes, so I was wondering the best way to do this. I would either inject it on components so it can be called in the lg-google-map component, or make another action on the controller and pass this down to the component. Is it possible to pass route actions to a component or only controller actions? Anyway let me know if a) you think this is necessary and b) how you would do it.

bantic replied:
You could likely use an initializer to inject it onto components, or your component could import the [lazy loading function](https://github.com/huafu/ember-google-map/blob/6257f5febe1bfa0a6776c7ffd07e4fcc9e13258f/addon/utils/load-google-map.js) and do it itself.

Here's the addon's initializer: https://github.com/huafu/ember-google-map/blob/6257f5febe1bfa0a6776c7ffd07e4fcc9e13258f/app/initializers/ember-google-map.js
You could make your own that did roughly the same thing. The import path would be the same.

There's nothing stopping you from putting the `loadGoogleMap` function into the controller's scope, from which it could be passed into the component, too:

```
// route
this.controller.set('loadGoogleMap', this.loadGoogleMap);

// template
{{lg-google-map loadGoogleMap=loadGoogleMap}}

// in compnent
var loadGoogleMap = this.get('loadGoogleMap');
loadGoogleMap().then(function(){
  // loaded
});
```

Injecting it onto components yourself or importing the function in the component (`import loadGoogleMap from 'ember-google-map/utils/load-google-map';`) are better ways, though.

However, I think there is value in displaying the map even when not editing, because a user may want to edit the image set because the location is incorrect. If the location/map is not displayed by default, though, they wouldn't know that it is correct and wouldn't want to edit it. (The [google maps static image api](https://developers.google.com/maps/documentation/imageapis/) may be more appropriate for non-interactive map displays, though.)

Thanks for being cognizant of the bandwidth/loading considerations and addressing them here, though.

iezer replied:
Very interesting. Thanks for the explanation. I'll make the map visible all the time and make marker.isDraggable flag bound to isEditing so you can only move it in editing mode.

bantic commented on app/components/lg-google-map.js:
```diff
@@ -0,0 +1,44 @@
+  isVisible: Ember.computed.alias('showMaps'),
+
+  showMaps: function() {
+    return !window.SKIP_MAPS;
```
this should be a config setting instead of stuck on the `window` global

bantic commented on app/components/lg-google-map.js:
```diff
@@ -0,0 +1,44 @@
+
+  markerLatObserver: function(){
+    var lat = this.get('markers.firstObject.lat');
+    this.set('marketLat', lat);
```
what is `marketLat`?

bantic commented on app/routes/image-sets/new.js:
```diff
@@ -1,12 +1,19 @@
     return this.store.createRecord('image-set', {
-      user: currentUser
+      user: currentUser,
+      latitude: Ember.get(defaultLocation, 'lat'),
```
it would be safe to do `defaultLocation.lat` since we know and can control what `defaultLocation` is. `Ember.get(object, propertyPath)` is mostly for cases where we do not control whether an `object` is an Ember Object or a pojo (i.e., if it were the input to a function). No harm in `Ember.get`-ing it here, but not necessary either.

bantic replied:
actually, now that I think of it, it may be most appropriate to make this a default value in the model definition for an image set. In image-set.js:

```
...
latitude: DS.attr('number', {defaultValue: defaultLocation.latitude}),
```

Also, fwiw, I would prefer to use the same property name everywhere (`latitude` instead of `lat`), for consistency. It's a little unexpected that the default latitude value for the image set is a property called 'lat' instead of 'latitude'.
Longitude is more error-prone because some people abbreviate it 'long' and others 'lng'.

bantic commented on app/models/image-set-marker.js:
```diff
@@ -0,0 +1,9 @@
+
+export default Ember.Object.extend({
+  imageSet: null,
+  lat: Ember.computed.alias('imageSet.latitude'),
```
instead of relying on two-way computed properties to update `imageSet.latitude` and longitude, the ember 2.0 way to do this would be to use an action. The `lg-google-map` component would be given an action and the observers for marker lat and lng would just fire that action with an updated [lat,lng] tuple. Bonus points for using debounce and an config'd timeout to avoid overloading things.

iezer replied:
I think a better solution is like we had in gwatt settings, where we have intermediate variables selectedLatitude, selectedAge, etc. so we set them when the user presses save, and they also have the option to cancel.

bantic commented on app/routes/image-set.js:
```diff
@@ -3,6 +3,17 @@ import OrganizationRouteMixin from 'lion-guardians/mixins/organization-route';
+    var route = this;
+    return this._super().then(function() {
+      if (window.SKIP_MAPS) {
+        return Ember.RSVP.resolve();
```
you do not need to return a resolved promise inside this function. You are already in "promise land" at this point and you only need to return if you are doing further async (like loading a map, or making an ajax call)

bantic replied:
to clarify: you do not need to return `Ember.RSVP.resolve()` here -- you can do nothing if you are not loading the map. You _do_ need to return `route.loadGoogleMaps()` if you are loading the map, though.

bantic commented:
looking pretty good! A few things to address that I mentioned above. Please ping me again after you've responded and I'll give this another look-over. Thanks @iezer !

iezer committed:
Image Set Editor can reset image set values

iezer commented:
@bantic  Please give this another look. I addressed your comments and made a few changes
- The map is always there but the marker is only draggable in edit mode
- The user can press cancel and all image set properties including marker position reset
- A 'Saving In Progress' message is shown while saving.

Here's what it looks like

![](http://g.recordit.co/wnic0vgbXX.gif)

bantic commented on app/models/image-set-marker.js:
```diff
@@ -0,0 +1,11 @@
+export default Ember.Object.extend({
+  component: null,
+  // these have to be called lat and lng to match
+  // expectations of google-map component
```
:+1: thanks for the comment

bantic commented on config/environment.js:
```diff
@@ -6,6 +6,11 @@ module.exports = function(environment) {
     locationType: 'auto',
+    googleMap: {
+      lazyLoad: true,
+      apiKey: 'AIzaSyCoc_T6TNxu5Mnw28aJleIm63fR7WzViUU'
```
I believe we'll need to get a key for the production domain as well, otherwise google maps will pop up an alert when this is deployed

bantic commented:
this is great. The animated gif screenshots are super helpful, please keep including them on your PRs.

I would think that we'd probably want some sort of visual indicator next to the map to indicate that the pin is draggable when the image set is editable, but that can happen separately.

thanks! If I'm correct about needing a second apiKey for the deployment domain can you generate that and add to the config? Otherwise this LGTM. Thank you

iezer commented:
@bantic  Right now I have 1 key configured that accepts users from both `localhost:4200` and `lion-guardians.appspot.com`  I released and it's working but I agree we should probably have separate key. I'll set that up later.
